nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py

```diff
@@ -134,91 +137,152 @@ def __getitem__(self, idx):
                 idx = idx.item()
 
         assert idx < len(self.indexed_dataset)
+        # idx may < 0 because we pad_samples_to_global_batch_size, e.g. id = -1
+        if idx < 0:
+            idx = len(self) + idx
         example = self.indexed_dataset[idx]
         return self._process_example(example)
 
+    def _calculate_total_ids(self, example):
```

Comment:
doctring

Reply:
this function duplicates multiple lines from process_example(), why is it needed?

Reply:
This is used to calculate how long we'll get for total tokens, so we can know how long we can truncate contexts.

Reply:
I try to refactor and only remain two functions `_process_prompt()` and `_process_truncation()`